pkg/storage/engine/temp_engine.go

```diff
@@ -62,52 +93,102 @@ type tempEngine struct {
 
 func (e *tempEngine) Close() {
 	e.RocksDB.Close()
+	// We could call cleanupTempStoreDirs here, but since we prefer faster
+	// shutdowns (whereas node start-up allow more time for us to cleanup
+	// asynchronously) we simply remove our current temporary store.
 	dir := e.RocksDB.cfg.Dir
 	if dir == "" {
 		return
 	}
 	if err := os.RemoveAll(dir); err != nil {
-		log.Errorf(context.TODO(), "could not remove rocksdb dir: %v", err)
+		log.Errorf(context.TODO(), "could not remove temporary store directory: %v", err.Error())
 	}
 }
 
-// wg is allowed to be nil, if the caller does not want to wait on the cleanup.
-func cleanupTempStorageDirs(ctx context.Context, path string, wg *sync.WaitGroup) error {
-	// Removing existing contents might be slow. Instead we rename it to a new
-	// name, and spawn a goroutine to clean it up asynchronously.
-	if err := os.MkdirAll(path, 0755); err != nil {
-		return err
-	}
-	deletionDir, err := ioutil.TempDir(path, "TO-DELETE-")
+// logTempStoreDir records the location of the temporary directory in a log
+// file for cleanup on startup if necessary.
+func logTempStoreDir(storePath string, tempPath string) error {
+	return logTempStoreDirs(storePath, append([]byte(tempPath), '\n'))
+}
+
+// logTempStoreDirs records tempPaths (list of temporary directory paths) to a log file
+// file for cleanup on startup if necessary.
+// tempPaths is a \n separated and ending list of paths to the temporary store directories.
+func logTempStoreDirs(storePath string, tempPaths []byte) error {
+	mu.Lock()
+	defer mu.Unlock()
+	// If the file does not exist, create it, or append to the file.
+	f, err := os.OpenFile(tempStoreDirsLogFilepath(storePath), os.O_APPEND|os.O_CREATE|os.O_WRONLY, 0644)
 	if err != nil {
 		return err
 	}
+	defer func() error {
+		if err := f.Close(); err != nil {
+			return err
+		}
+		return nil
+	}()
 
-	filesToDelete, err := ioutil.ReadDir(path)
+	// Append the path to the new temporary directory to the log file.
+	_, err = f.Write(tempPaths)
+	return err
+}
+
+// cleanupTempStoreDirs is invoked on startup to clean up abandoned temp
+// store directories from previous startups. It reads from the temporary store
+// log file to figure out which temporary stores need to be cleaned up.
+// wg is allowed to be nil, if the caller does not want to wait on the cleanup.
+// Removing existing temporary directories might be slow.  We spawn a
+// goroutine to clean it up asynchronously.
+func cleanupTempStoreDirs(ctx context.Context, storePath string, wg *sync.WaitGroup) error {
```

Comment:
This does the bulk of the heavy lifting for cleaning up old temporary stores during startup. Would love some feedback on this approach.